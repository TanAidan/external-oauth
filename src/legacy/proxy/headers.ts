const hopByHopHeaders = [
  'keep-alive',
  'proxy-authenticate',
  'proxy-authorization',
  'te',
  'trailer',
  'transfer-encoding',
  'upgrade',
  'host',
  'authorization',
  'x-forwarded-for' // to avoid sending the origin IP to the provider (because the rate limiting)
]

const connection = 'connection'
const BEARER_PREFIX = 'bearer-'
const AMAZON_PREFIX = 'x-amzn'

export const stripHopByHopHeaders = function(requestHeaders: any) {
  const requestHeadersCopy = Object.assign({}, requestHeaders)
  hopByHopHeaders.forEach((hopByHopHeaderName: string) => {
    delete requestHeadersCopy[hopByHopHeaderName]
  })
  stripConnectionHeaders(requestHeadersCopy)
  stripHttp2AndBearerHeaders(requestHeadersCopy)
  return requestHeadersCopy
}

const stripHttp2AndBearerHeaders = function(requestHeaders: any) {
  Object.keys(requestHeaders).forEach((headerName: string) => {
    if (headerName.includes(':') || headerName.startsWith(BEARER_PREFIX) || headerName.startsWith(AMAZON_PREFIX)) {
      delete requestHeaders[headerName.trim()]
    }
  })
}

const stripConnectionHeaders = function(requestHeaders: any) {
  let connectionHeaders = requestHeaders[connection]
  if (connectionHeaders) {
    connectionHeaders = lowercaseHeaders(connectionHeaders.split(','))
    delete requestHeaders[connection]
    connectionHeaders.forEach((connectionHeader: string) => {
      delete requestHeaders[connectionHeader.trim()]
    })
  }

  return requestHeaders
}

const lowercaseHeaders = function(headers: string[]): string[] {
  return headers.map((header: string) => {
    return header.toLowerCase()
  })
}
