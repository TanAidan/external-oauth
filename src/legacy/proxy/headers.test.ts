import { stripHopByHopHeaders } from './headers'

describe('Filter connection headers', () => {
  it('removes all hop by hop headers', () => {
    const headers = {
      te: 'gzip',
      host: 'example.net',
      'content-Type': 'application/json',
      connection: 'keep-alive, foo, bar',
      foo: 'abc',
      'keep-alive': 'timeout=30'
    }

    expect(stripHopByHopHeaders(headers)).toMatchSnapshot()
  })

  it('removes http2 headers', () => {
    const headers = {
      te: 'gzip',
      host: 'example.net',
      'content-Type': 'application/json',
      connection: 'keep-alive, Foo, Bar',
      foo: 'abc',
      'keep-alive': 'timeout=30',
      ':method': 'GET',
      ':scheme': 'https'
    }

    expect(stripHopByHopHeaders(headers)).toMatchSnapshot()
  })
})
