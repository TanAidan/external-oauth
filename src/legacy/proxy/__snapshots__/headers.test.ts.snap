// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter connection headers removes all hop by hop headers 1`] = `
Object {
  "content-Type": "application/json",
}
`;

exports[`Filter connection headers removes http2 headers 1`] = `
Object {
  "content-Type": "application/json",
}
`;
